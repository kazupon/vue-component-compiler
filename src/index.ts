import {
  SFCCompiler,
  StyleOptions,
  TemplateOptions,
  ScriptOptions,
  CustomBlockOptions
} from './compiler'

export const createCompiler = ({
  script,
  style,
  template,
  customBlock
}: {
  script: ScriptOptions
  style: StyleOptions
  template: TemplateOptions
  customBlock: CustomBlockOptions
}) => new SFCCompiler(script, style, template, customBlock)

export const createDefaultCompiler = (options: {
  script?: ScriptOptions
  style?: StyleOptions
  template?: TemplateOptions
  customBlock?: CustomBlockOptions
} = {}) =>
  createCompiler({
    script: { ...options.script },
    style: { trim: true, ...options.style },
    template: {
      compiler: require('vue-template-compiler'),
      compilerOptions: {},
      isProduction: process.env.NODE_ENV === 'production',
      optimizeSSR: process.env.VUE_ENV === 'server',
      ...options.template
    },
    customBlock: { ...options.customBlock }
  })

export * from './compiler'
export * from './assembler'
