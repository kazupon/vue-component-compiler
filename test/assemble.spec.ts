import {assembleFromSource, createDefaultCompiler} from "../src"

const compiler = createDefaultCompiler()
it('should inline style source', () => {
  const result = assembleFromSource(compiler, {}, {
    filename: 'foo.vue', script: {source: ''}, template: {source: ''}, scopeId: 'foo',
    styles: [{source: '.foo_0{ color: red }', module: { foo: 'foo_0' }, moduleName: '$style', map: {} }],
    customBlocks: []
  })

  expect(result.code).toEqual(expect.stringContaining('source: ".foo_0{ color: red }"'))
  expect(result.code).toEqual(expect.stringContaining('map: {}'))
  expect(result.code).toEqual(expect.stringContaining('value: {"foo":"foo_0"}'))
})

it('should use global in style source', () => {
  const result = assembleFromSource(compiler, {}, {
    filename: 'foo.vue', script: {source: ''}, template: {source: ''}, scopeId: 'foo',
    styles: [{source: 'fooStyle', module: 'fooModule', moduleName: '$style', map: 'fooMap'}],
    customBlocks: []
  })

  expect(result.code).toEqual(expect.stringContaining('source: fooStyle'))
  expect(result.code).toEqual(expect.stringContaining('map: fooMap'))
  expect(result.code).toEqual(expect.stringContaining('value: fooModule'))
})

it('should included source of custom blocks', () => {
  const result = assembleFromSource(compiler, {}, {
    filename: 'foo.vue', script: {source: ''}, template: {source: ''}, scopeId: 'foo',
    styles: [],
    customBlocks: [{
      source: [
        `export * from 'foo.vue?rollup-plugin-vue=customBlocks.0.i18n'`,
        `  import block0 from 'foo.vue?rollup-plugin-vue=customBlocks.0.i18n'`,
        `  if (typeof block0 === 'function') block0(Component)`
      ].join('\n'),
      type: 'i18n'
    }, {
      source: [
        `  export * from 'foo.vue?rollup-plugin-vue=customBlocks.1.json'`,
        `  import block1 from 'foo.vue?rollup-plugin-vue=customBlocks.1.json'`,
        `  if (typeof block1 === 'function') block1(Component)`
      ].join('\n'),
      type: 'i18n',
      lang: 'json'
    }]
  })
  expect(result.code).toMatchSnapshot()
})