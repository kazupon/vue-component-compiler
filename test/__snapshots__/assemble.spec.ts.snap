// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should included source of custom blocks 1`] = `
"/* script */

/* template */

  /* style */
  const __vue_inject_styles__ = undefined
  /* scoped */
  const __vue_scope_id__ = undefined
  /* module identifier */
  const __vue_module_identifier__ = undefined
  /* functional template */
  const __vue_is_functional_template__ = undefined
  /* component normalizer */
  function __vue_normalize__(
    template, style, script,
    scope, functional, moduleIdentifier, shadowMode,
    createInjector, createInjectorSSR, createInjectorShadow
  ) {
    const component = (typeof script === 'function' ? script.options : script) || {}

    // For security concerns, we use only base name in production mode.
    component.__file = \\"foo.vue\\"

    if (!component.render) {
      component.render = template.render
      component.staticRenderFns = template.staticRenderFns
      component._compiled = true

      if (functional) component.functional = true
    }

    component._scopeId = scope

    if (false) {
      let hook
      if (false) {
        // In SSR.
        hook = function(context) {
          // 2.3 injection
          context =
            context || // cached call
            (this.$vnode && this.$vnode.ssrContext) || // stateful
            (this.parent && this.parent.$vnode && this.parent.$vnode.ssrContext) // functional
          // 2.2 with runInNewContext: true
          if (!context && typeof __VUE_SSR_CONTEXT__ !== 'undefined') {
            context = __VUE_SSR_CONTEXT__
          }
          // inject component styles
          if (style) {
            style.call(this, createInjectorSSR(context))
          }
          // register component module identifier for async chunk inference
          if (context && context._registeredComponents) {
            context._registeredComponents.add(moduleIdentifier)
          }
        }
        // used by ssr in case component is cached and beforeCreate
        // never gets called
        component._ssrRegister = hook
      }
      else if (style) {
        hook = shadowMode 
          ? function(context) {
              style.call(this, createInjectorShadow(context, this.$root.$options.shadowRoot))
            }
          : function(context) {
              style.call(this, createInjector(context))
            }
      }

      if (hook !== undefined) {
        if (component.functional) {
          // register for functional component in vue file
          const originalRender = component.render
          component.render = function renderWithStyleInjection(h, context) {
            hook.call(context)
            return originalRender(h, context)
          }
        } else {
          // inject component registration as beforeCreate hook
          const existing = component.beforeCreate
          component.beforeCreate = existing ? [].concat(existing, hook) : [hook]
        }
      }
    }

    return component
  }
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const Component = __vue_normalize__(
    {},
    __vue_inject_styles__,
    {},
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    undefined,
    undefined,
    undefined
  )
  
  /* custom blocks */
  import block0 from 'foo.vue?rollup-plugin-vue=customBlocks.0.i18n'
  if (typeof block0 === 'function') block0(Component)
  import block1 from 'foo.vue?rollup-plugin-vue=customBlocks.1.json'
  if (typeof block1 === 'function') block1(Component)

  export default Component"
`;
